File: src/emire/extract.py
import re


def extract_json(markdown: str) -> str:
    r"""
    >>> md = '''
    ... Here is some text.
    ... ```json
    ... {
    ...   "key": "value",
    ...   "number": 123
    ... }
    ... ```
    ... More text here.
    ... '''
    >>> extract_json(md)
    '{\n  "key": "value",\n  "number": 123\n}'
    >>> import json
    >>> json.loads(extract_json(md))
    {'key': 'value', 'number': 123}
    """
    pattern = r"```json\s*(.*?)\s*```"
    match = re.search(pattern, markdown, re.DOTALL)
    if match:
        return match.group(1)

    message = "No JSON block found in the provided markdown."
    raise ValueError(message)

File: emire-core/src/extract.rs
use regex::Regex;

pub fn extract_json(markdown: &str) -> Option<&str> {
    let re = Regex::new(r"(?s)```json\s*(.*?)\s*```").unwrap();
    let captures = re.captures(markdown)?;
    Some(captures.get(1)?.as_str())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_extract_json() {
        let markdown = r#"
Here is some text.
```json
{
  "key": "value",
  "number": 123
}
```
More text here.
"#;
        assert_eq!(
            extract_json(markdown),
            Some(
                r#"{
  "key": "value",
  "number": 123
}"#
            )
        )
    }
}
